design_documents/prepare_cheatcode.md

```diff
@@ -0,0 +1,122 @@
+
+## Goal
+
+Propose a solution that will allow knowing the address of a contract before the deployment.
+
+## Considered Solutions
+
+### Require Calling the `prepare` Cheatcode Before Every Deployment
+
+This is similar to how the contract deployment worked in Protostar:
+
+1. Call `declare` with the contract name. This returns `class_hash`.
+2. Call `prepare` with the `class_hash` and `calldata`. This returns `PreparedContract` struct.
+3. Call `deploy` with the `PreparedContract` struct. This returns the address.
+
+With this approach, `PreparedContract` also included a `contract_address` field that could be used for applying the
+cheatcodes before the contract was deployed.
+
+The problem with this approach is that it is very verbose and requires multiple steps:
+Just deploying the contracts is more frequent use case than applying cheatcodes before the deployment.
+Users would have to perform an often unnecessary extra step with every deployment.
+
+### Introducing the `precalculate_address` Cheatcode
+
+Introducing the `precalculate_address` cheatcode that would return the contract address of the contract that would be
+deployed with `deploy` cheatcode.
+The problem with this approach is that for it to work, contract addresses would have to be deterministic.
+This would limit the user to only deploying a one instance of the given contact.
+
+## Proposed Solution
+
+Improving on the solution [proposed here](#introducing-the-precalculateaddress-cheatcode), allow precalculating the
+address while having the contract address semi-deterministic.
+
+### Salt "Counter"
+
+Introduce an internal "counter" and use its value to salt the otherwise deterministic contract address.
+Every time the `deploy` is called, increment this counter, so subsequent calls of `deploy` with the same `class_hash`
+will yield different addresses.
+
+### `precalculate_address` Cheatcode
+
+Introduce a cheatcode with the signature:
+
+```cairo
+fn precalculate_address(prepared_contract: PreparedContract) -> ContractAddress;
+```
+
+That will use the same method of calculating the contract address as `deploy` uses, utilizing the internal counter.
+This way the user will have an ability to know the address of the contract that will be deployed, and the current
+deployment flow will remain unchanged.
+
+### Known Problems With This Solution
+
+For the address returned by `precalculate_address` to match the address from `deploy`, the user will have to
+call `precalculate_address` immediately before the deployment or at least before any other calls to `deploy` as the
+internal counter will then be incremented.
+
+This could be remedied by having separate counters for all `class_hashe`es, but it will still remain a limiting factor.
+
+## Example Usage
+
+```cairo
+mod HelloStarknet {
+    // ...
+    
+    #[constructor]
+    fn constructor(ref self: ContractState) {
+        let caller_address = starknet::get_caller_address();
+        self.owner.write(caller_address);
+    }
```

Comment:
Have been discussing with @Radinyn about this prank use case lately.

This approach to saving owner address is one that you would almost never use on real network, since most of the contracts are deployed through UDC (unless you have an entrypoint calling deploy syscall in your contract address) and `get_caller_address` would return UDC's address. I think we shouldn't use this example in this design doc (and not replicate it anywhere anymore, maybe even state that it is a bad practice and it won't work in real starknet most of the times?), because in real network there is no way you can prank UDC's address.

Reply:
cc @MaksymilianDemitraszek @war-in

Reply:
Changing this example so it uses `warp` instead.